### ISSUE TITLE
Voila with jupyterlab and jupyterhub

### ISSUE THREAD

**Issue by anki-code** [voila-dashboards/voila#437]
Hi! Thank you for great project!

I have jupyterlab behind jupyterhub. It's one docker container and users have access to terminal. Is there a way to allow users create voila dashboards and share it 1. inside jupyterhub 2. outside jupyterhub?

Thanks!

**Comment by jtpio** [comment]
If everything is in one Docker container then there could be a separate Voila process running in the container that serves the dashboards. And that would be outside JupyterHub.

We're still working on native Voila support within JupyterHub (for example being able to run Voila as a named server per user, or having a Voila service for JupyterHub). In the meantime users can still use the server extension endpoint.

**Comment by jtpio** [comment]
I was going to reference https://github.com/voila-dashboards/voila/issues/112 but it seems like you posted the last message :)

**Comment by davhin** [comment]
> If everything is in one Docker container then there could be a separate Voila process running in the container that serves the dashboards. And that would be outside JupyterHub.
> 
> We're still working on native Voila support within JupyterHub (for example being able to run Voila as a named server per user, or having a Voila service for JupyterHub). In the meantime users can still use the server extension endpoint.

What do you mean by that endpoint? If I run voila inside my bash on jupyterlab it seems to work ok, but nothing happens.

**Comment by jtpio** [comment]
The endpoint that is added to the notebook server (and should be enabled automatically on installed).

So that JupyterLab is available with `/lab` and Voila with `/voila`: https://voila.readthedocs.io/en/latest/using.html#as-a-jupyter-server-extension

**Comment by anki-code** [comment]
@jtpio great thanks! 

Is there a way to set restrictions? For example there are 3 data scientists who use `/lab` and there are many users who want to review notebooks in folder `shared` from `/voila` path. Data scientists place the notebooks to `shared` folder and send the links to users.

I already use [outhenticator.google](https://github.com/jupyterhub/oauthenticator) to authorize a few data scientists in the `/lab`. Can I set different users list for `/voila`?

**Comment by jtpio** [comment]
@anki-code Do the data Scientists and users share the same hub? One option could be to implement a custom `Spawner` and implement the logic there.

**Comment by anki-code** [comment]
@jtpio yep. Data scientist just make simple notebook for results and want to share it via voila. It will be great if data user can see only the notebooks were shared and pass thru authorization for that.

**Comment by dschofield** [comment]
@jtpio something similar to the way [Voila Gallery ](https://voila-gallery.org/) presents and runs notebooks with Voila (with authentication prior) would be ideal. Any setup and configuration details from Voila Gallery would help, if you can share those.

**Comment by dschofield** [comment]
Referring to https://github.com/wildtreetech/ohjh/blob/master/images/user/Dockerfile#L29 I see:
`RUN echo "c.ServerProxy.servers = {" >> /etc/jupyter/jupyter_notebook_config.py
RUN echo "'voila': {" >> /etc/jupyter/jupyter_notebook_config.py
RUN echo "'command': ['voila', '--port', '{port}', '--server_url', '/', '--base_url', '{base_url}/voila/']}}" >> /etc/jupyter/jupyter_notebook_config.py
`
which goes a long way to show how to do this.

**Comment by jtpio** [comment]
@dschofield the configuration for running the gallery as a plugin for The Littlest JupyterHub (TLJH) is in this repo: https://github.com/voila-dashboards/gallery

This is the same configuration as the one used by the public-facing gallery. But it still spawns one notebook server per user, with the voila server extension enabled.

> Checking the user info to determine whether a regular single-user server vs. a voila-ified one should be launched seems straightforward enough. I'm a bit unclear as to what exactly the spawner launch cmd needs to look like to achieve voila-ificiation.

Starting one instance running voila standalone per user, not one single instance running voila standalone shared by all the users?

**Comment by danlester** [comment]
My [ContainDS Dashboards](https://github.com/ideonate/cdsdashboards) project sounds perfect here. 

It is an extension for JupyterHub that allows users to instantly and securely clone a Jupyter notebook into a separate Voila server, accessible to other authenticated users of JupyterHub.

Non-technical users can interact with the Voila front-end without having to step through code, or being able to switch to the /tree URL to expose the regular Jupyter development environment.

Instructions are here: https://cdsdashboards.readthedocs.io/en/stable/index.html

At the moment, the default is for sharing internally within JupyterHub, but it would be possible to configure it to display publicly. Please get in touch if you would like to discuss further or need any help.
